rust/src/ike/detect.rs

```diff
@@ -0,0 +1,263 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// Author: Frank Honza <frank.honza@dcso.de>
+
+use crate::ike::ike::*;
+use std::ptr;
+use std::ffi::CStr;
+use super::ipsec_parser::IkeV2Transform;
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_exch_type(
+    tx: &mut IKETransaction,
+    exch_type: *mut u8,
+) -> u8 {
+    if exch_type == std::ptr::null_mut() {
+        return 0;
+    }
+
+    if tx.ike_version == 1 {
+        if let Some(r) = tx.hdr.ikev1_header.exchange_type {
+            unsafe {
+                *exch_type = r;
+            }
+            return 1;
+        }
+    } else if tx.ike_version == 2 {
+        unsafe {
+            *exch_type = tx.hdr.ikev2_header.exch_type.0;
+        }
+        return 1;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_spi_initiator(
+    tx: &mut IKETransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut() {
+        return 0;
+    }
+
+    unsafe {
+        *buffer = tx.hdr.spi_initiator.as_ptr();
+        *buffer_len = tx.hdr.spi_initiator.len() as u32;
+    }
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_spi_responder(
+    tx: &mut IKETransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut() {
+        return 0;
+    }
+
+    unsafe {
+        *buffer = tx.hdr.spi_responder.as_ptr();
+        *buffer_len = tx.hdr.spi_responder.len() as u32;
+    }
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_nonce(
+    tx: &mut IKETransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    if buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut() {
+        return 0;
+    }
+
+    if !tx.hdr.ikev1_header.nonce.is_empty() {
```

Comment:
Should we not first check we have ike _v1_ here ?

Reply:
Not sure if needed, the data structure is only filled if we have a v1 communication. Perhaps reasonable if integer comparision faster than check for empty string? (not sure here)

Reply:
> the data structure is only filled if we have a v1 communication

 Do not we risk some kind of null dereference then ?

That seems reasonable to me to add an integer comparison...

Reply:
ok